src/Handler/StreamHandler.php

```diff
@@ -301,21 +302,44 @@ private function createStream(RequestInterface $request, array $options)
             );
         }
 
+        $uri = $this->resolveUri($request->getUri(), $options);
+
         $context = $this->createResource(
             function () use ($context, $params) {
                 return stream_context_create($context, $params);
             }
         );
 
         return $this->createResource(
-            function () use ($request, &$http_response_header, $context) {
-                $resource = fopen((string) $request->getUri()->withFragment(''), 'r', null, $context);
+            function () use ($uri, &$http_response_header, $context) {
+                $resource = fopen((string) $uri, 'r', null, $context);
                 $this->lastHeaders = $http_response_header;
                 return $resource;
             }
         );
     }
 
+    private function resolveUri(UriInterface $uri, array $options)
+    {
+        if (isset($options['force_ip_resolve'])) {
+            if ('v4' === $options['force_ip_resolve']) {
+                $records = dns_get_record($uri->getHost(), DNS_A);
+                if (!isset($records[0]['ip'])) {
+                    throw new ConnectException("Could not resolve IPv4 address for host '{$uri->getHost()}'", $request);
```

Comment:
Can you please use sprintf instead? I am not really fan of this kind of resolution.